lib/btree.js

```diff
@@ -0,0 +1,115 @@
+    this.less = less;
+  }
+}
+
+const empty = less => new Element(undefined, null, less);
+
+const destroyNode = (parent, index) => {
+  const node = parent[index].less;
+  const len = node.length;
+  const newLeftNode = node.splice(0, len / 2 - 1); // First half
+  const mid = node.shift();
+  newLeftNode.push(empty(mid.less));
+  mid.less = newLeftNode;
+  parent.splice(index, 0, mid);
+  return parent;
+};
+
+const isLeaf = node => !node[0].less;
+
+const binarySearch = (node, value) => {
+  let start = 0;
+  let end = node.length - 1;
+  while (start <= end) {
+    const i = (start + end) >> 1;
+    const item = node[i];
+    const itemValue = item.value;
+    if (value > itemValue) start = i + 1;
+    else if (!(value >= itemValue)) end = i - 1;
+    // undefined must be bigger then other value
+    // (!(1 >= undefined)) -> true, but just (1 < undefined) -> false
+    else return [true, item.tuple];
+  }
+  return [false, start];
+};
+
+function* inorderTraversal(start, finish, currNode) {
+  let [found, startIndex] = binarySearch(currNode, start);
+  if (found) startIndex++;
+  let [_, finishIndex] = binarySearch(currNode, finish);
+  if (finish === undefined)
+    finishIndex = currNode.length - 1;
+  for (let currIndex = startIndex; currIndex <= finishIndex; currIndex++) {
+    const currElement = currNode[currIndex];
+    if (!(isLeaf(currNode)))
+      yield* inorderTraversal(start, finish, currElement.less);
+    if (currIndex < finishIndex)
+      yield currNode[currIndex].tuple;
+  }
+
+}
+
+class BTree {
+
+  constructor(degree = DEFAULT_DEGREE) {
+    this.root = [empty()];
+    this.minDegree = degree;
+  }
+
+  get(value) {
+    let currNode = this.root;
+    while (currNode) {
+      const [found, i] = binarySearch(currNode, value);
+      if (found)
+        return i;
+      currNode = currNode[i].less;
+    }
+    return null;
+  }
+
+  insert(value, tuple) {
+    const alreadyHaveErrMes = `value ${value} already is in the tree`;
+    const newElement = new Element(value, tuple);
+    if (this.root.length === 1)
+      return (this.root.unshift(newElement), this);
+    if (this.root.length === this.minDegree * 2) {
+      this.root = [empty(this.root)];
+      destroyNode(this.root, 0);
+    }
+    let currNode = this.root;
+    while (true) {
+      const [found, nextNodeIndex] = binarySearch(currNode, value);
+      if (found)
+        throw new Error(alreadyHaveErrMes);
+      if (isLeaf(currNode))
+        return (currNode.splice(nextNodeIndex, 0, newElement), this);
+      let nextNode = currNode[nextNodeIndex].less;
+      if (nextNode.length === this.minDegree * 2) {
+        destroyNode(currNode, nextNodeIndex);
+        const element = currNode[nextNodeIndex];
+        if (element.value === value)
+          throw new Error(alreadyHaveErrMes);
+        if (element.value > value)
+          nextNode = element.less;
+      }
+      currNode = nextNode;
+    }
+  }
+
+  iterator(start, finish = undefined) {
```

Comment:
`start` can also be undefined, you will then traverse the whole tree.

Reply:
I remembered that if we didn`t pass the value into the argument of the function, then it would become an ``undefined``. Therefore, we can write just this: -
``iterator(start, finish) {}``

Reply:
Yes, if you don't need specific values (not undefined) for defaults you can just remove the defaults.